Add "Open in new window" option to links

notnownikki opened the pull request:
Fixes https://github.com/WordPress/gutenberg/issues/2204

notnownikki committed:
Add "Open in new window" option to links

youknowriad commented on blocks/editable/format-toolbar/index.js:
```diff
@@ -130,6 +146,13 @@ class FormatToolbar extends Component {
 			} ) );
 
+		// TODO: make this not look hideous
+		const linkSettings = settingsVisible && (
```
Have you considered using the `Popover` component?

notnownikki replied:
Nope, mainly because I didn't know what it was :smile: Looks like a good way to have this visibility state managed for us, although I don't think it would help with my awful design skills?

notnownikki replied:
I've switched this to use a popover, but it still looks really bad. I'm not sure what markup we need to have this render like the wireframes, and there's nowhere that we're actually using a PopoverContext for me to see how to get this rendering nicely, either. Also, there are focus issues once you toggle the checkbox (click the settings icon, toggle the checkbox, then you have to click the settings icon twice to close the Popover, as I think the first click takes the focus away from the Popover, which feels bad.)

notnownikki committed:
Works, but still looks awful

notnownikki committed:
lint

notnownikki committed:
Remove popover, use ToggleControl

notnownikki commented:
This works now, and is using the correct form components, but for the life of me I can't get the settings to appear below the toolbar, they just squish everything up. Perhaps @karmatosed can work some magic here, or advise me?

mtias committed:
Display link-settings below the input and button tools.

mtias commented:
Tweaked the styling a bit:

![image](https://user-images.githubusercontent.com/548849/29970906-c10dcb4e-8f26-11e7-8e70-d83b78ae127e.png)

notnownikki committed:
Remove TODO, as it no longer looks hideous

notnownikki committed:
Set aria-expanded correctly

notnownikki commented:
I've set `aria-expanded` correctly now, and made sure that keyboard navigation works as expected. Here's what it looks like expanding the expanded settings using the keyboard:

![keyboard nav for link settings](https://user-images.githubusercontent.com/3314354/30019174-863b0972-9157-11e7-8dc0-66f4b48017aa.gif)

Perhaps @afercia could make sure I've got this right? Happy to squash and merge if everything looks ok :)

youknowriad commented on blocks/editable/format-toolbar/index.js:
```diff
@@ -93,6 +99,17 @@ class FormatToolbar extends Component {
 
+	toggleLinkSettingsVisibility() {
+		const settingsVisible = ! this.state.settingsVisible;
+		this.setState( { settingsVisible } );
```
there's an implicit coding standard stating that if the next state depends on the previous, we should use the callback notation (I think we were trying to add an eslint for this in an old PR)

```js
this.setState( ( state ) => ( { settingsVisible: ! state.settingsVisible } ) );
```

notnownikki replied:
Fixed!

youknowriad reviewed (commented):
Should we leave the settings visible when we toggle the target? They are being closed in my testing

youknowriad commented:
Am I the only one seeing this design bug

<img width="313" alt="screen shot 2017-09-04 at 13 13 26" src="https://user-images.githubusercontent.com/272444/30026071-3edc9c3c-9173-11e7-91f9-853073ca37c7.png">

notnownikki commented:
@youknowriad I'm not seeing either of those things happen... what browser/OS?

notnownikki commented:
(that is, setting stay open when I toggle the setting, and the toolbar doesn't jump to the next line)

youknowriad commented:
Weird!!! MacOS Chrome Version 60.0.3112.113. I reproduce in Firefox too 55.0.3 (64 bits)

notnownikki commented:
I see the setting icon wrap onto the next line in firefox, but I can't reproduce the "settings close when setting is toggled" thing in  any browser.

I'll have to defer to someone else to fix the wrapping bug though, still not sure how to get that working correctly

notnownikki commented:
Hmm, perhaps there's a click event we need to stop bubbling up somewhere? I can't see where that would be though, the settings fieldset isn't inside the IconButton...

karmatosed commented:
The grumpy z-index monster is back:

![2017-09-04 at 15 14](https://user-images.githubusercontent.com/253067/30030126-c28bee60-9183-11e7-86cf-255dbb4f10db.png)

This only happens when the screen size is reduced.

karmatosed commented:
After testing I have found the same FF issue:

![2017-09-04 at 15 15](https://user-images.githubusercontent.com/253067/30030153-e7dd370a-9183-11e7-8527-3c501e59adc6.png)

However, I can't replicate the other one.

Design wise I think if we can work out those issues, I would love to get this into 1.1. I am not sure this is perfect, but I would like to get user feedback on it and that's a great way to do it.

notnownikki commented:
I'm not entirely sure how to fix the FF wrapping issue... if anyone has a bit of spare time, I'd be happy to learn from them :)

notnownikki committed:
Made the modal slightly wider to fix a wrap problem in FF

notnownikki committed:
Coding standard fix

notnownikki commented:
I think I've addressed the issues here, it renders fine in chrome and firefox now, although I can't reproduce the issue where the settings close when you toggle a setting, and I can't see the design issue @youknowriad saw where all the link icons where wrapped onto the next line, no matter what browser I use.

karmatosed commented:
Looks good to me, so going to merge.
